cpp/src/io/parquet/parquet.hpp

```diff
@@ -157,7 +161,8 @@ struct SchemaElement {
     return type == other.type && converted_type == other.converted_type &&
            type_length == other.type_length && repetition_type == other.repetition_type &&
            name == other.name && num_children == other.num_children &&
-           decimal_scale == other.decimal_scale && decimal_precision == other.decimal_precision;
+           decimal_scale == other.decimal_scale && decimal_precision == other.decimal_precision &&
+           field_id == other.field_id;
```

Comment:
You can store an optional member in this struct. Then the equality check will just work

Reply:
I was using an optional member in the first place and the leftover of including thrust optional at the top of this file can prove this :smile: . The problem is during [reading](
https://github.com/PointKernel/cudf/blob/parquet-field-id-writing/cpp/src/io/parquet/compact_protocol_reader.cpp#L159), where we need to read the value of the optional and the optional can have no value:
```cpp
bool CompactProtocolReader::read(SchemaElement* s)
{
  auto op = std::make_tuple(ParquetFieldEnum<Type>(1, s->type),
                        ParquetFieldInt32(2, s->type_length),
                        ParquetFieldEnum<FieldRepetitionType>(3, s->repetition_type),
                        ParquetFieldString(4, s->name),
                        ParquetFieldInt32(5, s->num_children),
                        ParquetFieldEnum<ConvertedType>(6, s->converted_type),
                        ParquetFieldInt32(7, s->decimal_scale),
                        ParquetFieldInt32(8, s->decimal_precision),
                        ParquetFieldInt32(9, s->field_id.value()), // read a nullable value
                        ParquetFieldStruct(10, s->logical_type));
  return function_builder(this, op);
}
```
We don't know whether `field_id` has been set or not beforehand so when `field_id` is not set, execution will throw:
```
C++ exception with description "Optional has no value" thrown in the test body.
```
I haven't figured out how to handle a `thrust::optional` with the current compact protocol reader. I guess a new optional field reader is required? By all means, I'm removing the changes to the compact protocol reader so it won't read `field_id` for now.

Reply:
I would be really nice to read the field_id values in the reader. I think we can add `ParquetFieldOptionalInt32` for this purpose, or even turn `ParquetFieldInt32` into a template so it can also hold a reference to an optional. I think implementation is the same.
This is the exact reason why I had to refactor the ORC equivalent and use templates/overloads.

Reply:
I was thinking to only use it for the field ID for now. It bugs me that we can't round trip an ID. I think the limited tests demonstrate that the feature is incomplete. 
Not blocking the PR on this, but would prefer to have minimal reader changes included.

Reply:
Got it. Will do.